### PR TITLE
Update programming-assistants.rst and programming-assistants.bib

### PR DESCRIPTION
**Pull request by abhi-ganta**

### CODE REVIEW

File: src/programming-assistants.rst [hunk 1]
```diff
@@ -1,23 +1,179 @@
-.. :Authors: - Cyrus Omar
+.. :Authors: - Cyrus Omar, Abhitej Ganta
 
 .. title:: Programming Assistants
 
 Overview
 ========
 
-TODO: Overview
+Development software has improved far beyond what was initially conceived. **Programming assistants**, tools that are integrated into this software to enable more effective programming, are leading the charge with new initiatives and improvements every year. Why do we need programming assistants? As more complex and time-consuming projects arise, assistants can significantly reduce the time and effort a programmer exhausts during development.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: - Drop the first sentence (too general)
- Start with "Programming assistants are ..." and be specific.
- Avoid vague words like "initiatives and improvements"
- Maybe list some benefits of programming assistants in a bulleted list rather than mentioning only "reducing the time and effort"

File: src/programming-assistants.rst [hunk 2]
```diff
@@ -1,23 +1,179 @@
...
+
+Programming assistants can be highly sophisticated themselves, with most using advanced artificial intelligence, machine learning and natural language processing models. However, most models can be represented by the following diagram:
+
+-- Insert Diagram Here --
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Can you include the diagram in your second draft?

File: src/programming-assistants.rst [hunk 3]
```diff
@@ -1,23 +1,179 @@
...
+
+where the input can be range from syntactical structure (abstract syntax trees, parsed text) to program runtime data. The chosen model is done so based on accuracy and runtime comparisons between various hypothesized models. Even after the model is chosen, it is further evaluated and tuned. Once a model is chosen and trained on input data, new data is introduced for prediction and the results are published. The main areas of research for programming assistants are discussed further below.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: - also mention edit action history as an input
- also enumerate what might be predicted as output (filling a single hole, filling multiple holes, arbitrary edits throughout a program)

File: src/programming-assistants.rst [hunk 4]
```diff
@@ -1,23 +1,179 @@
...
 
 Refactoring
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Move Program Synthesis up before Refactoring

File: src/programming-assistants.rst [hunk 5]
```diff
@@ -1,23 +1,179 @@
...
 ===========
 
+**Refactoring** refers to the process of restructuring existing code without changing the program's behavior and is a large area of interest in programming assistant research. It is intended to improve the nonfunctional attributes of a program including code readability, reduced complexity and source code maintainability. Typically, code refactoring applies small changes to various areas of a program and most development environments have already integrated support for basic refactoring.
+
+Refactoring is motivated by indications of characteristics in a program of a deeper problem. This is known as **code smell**. Detecting problems early and refactoring source code can save developers time and effort debugging in the future. The main benefits of refactoring can be categorized into either **maintainability**: Fixing issues is very easy because the code is easy to read and easy to understand, or **extensibility**: Extending the capabilities of the application is very easy because the codes uses recognizable design patterns and is flexible.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Not all refactorings are motivated by code smells. Maybe we can move discussion of code smells elsewhere in the compendium, e.g. in the Language Usability chapter?

Comment (cyrus-) [inline src/programming-assistants.rst]: Maintainability is mentioned above, and then again here. Extensibility is only mentioned here. Consider either merging the two or pulling out all discussion of the benefits of refactoring into this paragraph.

File: src/programming-assistants.rst [hunk 6]
```diff
@@ -1,23 +1,179 @@
...
+
+Software engineers at Microsoft are currently developing and integrating a refactoring assistant that recognizes editing patterns in source code and suggest other edits :cite:`miltner2019fly`.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Blue Pencil will suggest program edits that are not just refactorings, so it should probably be discussed in a separate section on "Suggesting Program Edits"...

there is other work on suggesting refactorings (but you can leave that as a TODO here for now)

File: src/programming-assistants.rst [hunk 7]
```diff
@@ -1,23 +1,179 @@
...
+
+Although this is a small example, making changes like these is a very tedious process. Scaling this up to larger projects creates a real problem with a large resource requirement. This is the exact problem the Blue-Pencil solution aims to tackle. Blue-Pencil non intrusively watches and learn from user edits and makes suggestions accordingly. It attempts to identify repetitive edits using the document edit history by generating explanations for each of the user's edits to predict actions to take. `This article <https://devblogs.microsoft.com/visualstudio/refactoring-made-easy-with-intellicode/>`_ demonstrates Blue-Pencil in action as it was recently integrated into Visual Studio 2019.
+
+How does it work? Given a history of edits on a given document, Blue-Pencil takes an oracle-guided approach to generate explanations for the history. The following are the three main ingredients in an OracleBasedExplanationGenerator:
+
+ 1. An oracle that returns a program that explains a set of non-overlapping edits
+ 2. An exhaustive enumeration based procedure for creating a directed acyclical multigraph (DAM) where the nodes represent document versions and the edges represent edges
```
Comment (cyrus-) [inline src/programming-assistants.rst]: "the edges represent edges"?

File: src/programming-assistants.rst [hunk 8]
```diff
@@ -1,23 +1,179 @@
...
+A new function is then written to initialize the created variables so the code must be changed to call the new function:
+.. code-block:: python
+  var1 = NewFunction(param1)
+  var2 = NewFunction(param2)
+  var3 = NewFunction(param3)
+  var4 = NewFunction(param4)
+
+Although this is a small example, making changes like these is a very tedious process. Scaling this up to larger projects creates a real problem with a large resource requirement. This is the exact problem the Blue-Pencil solution aims to tackle. Blue-Pencil non intrusively watches and learn from user edits and makes suggestions accordingly. It attempts to identify repetitive edits using the document edit history by generating explanations for each of the user's edits to predict actions to take. `This article <https://devblogs.microsoft.com/visualstudio/refactoring-made-easy-with-intellicode/>`_ demonstrates Blue-Pencil in action as it was recently integrated into Visual Studio 2019.
+
+How does it work? Given a history of edits on a given document, Blue-Pencil takes an oracle-guided approach to generate explanations for the history. The following are the three main ingredients in an OracleBasedExplanationGenerator:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: This might be too much implementation detail for the compendium...

File: src/programming-assistants.rst [hunk 9]
```diff
@@ -1,23 +1,179 @@
...
+ 
+The model was evaluated with 37 document editing sessions in two software development languages (C# and SQL) coming from various sources. The model provided 206 suggestions and only yielded 23 false positives giving it a final precision value of 0.89 with the average suggestion time being 199 ms. More details on the implementation and results of the study can be found here.
+
 Program Synthesis
 =================
 
+**Program Synthesis** refers to the process of code synthesis based on a programmer's instructions. It is intended to serve as the ultimate abstraction and allows the assistant to implement instructions containing a mixture of English and programmatic syntax. This area of research originates from the idea that computers are very good at following explicit instructions, but humans are not necessarily good at writing those instructions. Wouldn't it be more efficient to tell a computer what the developer wants it to do and to leave the implementation details to the computer? The main promise of Program Synthesis is to automate the trivial parts of programming, allowing the developer to focus on the big picture.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: - "programmer's instructions" => "programmer's specifications"
- not all program synthesis work is based  on English language specifications (this should be a sub-section of program synthesis)

File: src/programming-assistants.rst [hunk 10]
```diff
@@ -1,23 +1,179 @@
...
+
+Another Program Repair tool, SemFix is an automated program repair method based on symbolic execution, constraint solving and program synthesis. It utilizes statistical fault localization in order to identify and rank lines of code based on their suspiciousness, determines the correct specifications of buggy statements using a method similar to angelic debugging and finally uses program synthesis to correct the statement :cite:`nguyen2013semfix`.
+
+To evaluate the tool, a buggy test set of 50 was used with a total of 90 bugs. The performance and speed of SemFix was also compared to those of GenProg, a competing automatic debugging tool. SemFix proved to be more successful and faster with the average speed repair speed being 100 ms. Although SemFix outperformed GenProg, it was still only able to debug 48 of the 90 bugs successfully (GenProg was only able to debug 16 of the 90) which seems problematic. However, the SemFix team clearly addresses this and explains the drawbacks of the tool. More details can be found `here <https://www.comp.nus.edu.sg/~abhik/pdf/ICSE13-SEMFIX.pdf>`_.
+
 Interactive Proof Assistants
 ============================
+
+
+Bibliography
```
Comment (cyrus-) [inline src/programming-assistants.rst]: The idea was for these items to be integrated into the text, with a short summary of each paper (like `notation.rst`). Can you do that for the second draft?

File: src/programming-assistants.rst [hunk 11]
```diff
@@ -1,23 +1,179 @@
...
+
+There are four key aspects in this tool:
+1. Completion across multiple types - In one given run, the tool is able to make completions for objects of various types.
+2. Complete of parameters - While generating completions for methods, the tool is also able to fill in the correct method parameters.
+3. Holes as sequences - The tool is able to recognize if multiple method invocations are missing in a given hole and fills in the missing.
+4. New fused completions - The tool is able to make completions on sequences and methods it has not been trained on before.
+
+The evaluation of the tools proved it to be very fast and effective. Out of the 84 examples tested on, the desired completion appeared in the top 3 results in 90% of the cases. More details on the implementation of the model and the evaluation metrics used can be found `here <http://www.cs.technion.ac.il/~yahave/papers/pldi14-statistical.pdf>`_.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Instead of linking to the paper, you should cite it.

File: src/programming-assistants.rst [hunk 12]
```diff
@@ -1,23 +1,179 @@
...
+
+However, when both models were used, the expected expression was in the top 10 82% of time and was the #1 option 44% of the time. More implementation details of the two models can be found `here <https://lara.epfl.ch/~kuncak/papers/GveroKuncak15SynthesizingJavaExpressionsFreeFormQueries.pdf>`_. anyCode was given an average runtime of 60 ms to generate the top 10 expressions, proving that it can effectively and efficiently synthesize expressions.
+
 Code Completion
 ---------------
 
+**Code Completion**, a subsection of Program Synthesis, refers to the process of intelligent completion suggestion making for lines of code a user writes during the development of an application. It is intended to speed up the process of coding applications by reducing typos and other mistakes. Code Completion assistants serves as convenient way to access function syntax and descriptions, significantly reducing the amount of memorization required.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: This isn't the clearest definition of code completion. I'd say something about how code completion systems are simple program synthesis systems that generate code snippets for insertion at the cursor (whereas other program synthesis systems might fill multiple holes).

Comment (cyrus-) [reply src/programming-assistants.rst]: By that definition, though, the Raychev 2014 system isn't code completion.

It might be better to just define "code completion system" as a phrase that means "program synthesis integrated into the editor".

File: src/programming-assistants.rst [hunk 13]
```diff
@@ -1,23 +1,179 @@
...
+
 Code Completion
 ---------------
 
+**Code Completion**, a subsection of Program Synthesis, refers to the process of intelligent completion suggestion making for lines of code a user writes during the development of an application. It is intended to speed up the process of coding applications by reducing typos and other mistakes. Code Completion assistants serves as convenient way to access function syntax and descriptions, significantly reducing the amount of memorization required. 
+
+Common techniques to achieve this include auto completion popups while typing, querying parameters of functions, querying hints related to syntax errors. The models used in Code Completion are very similar to those used in Program Synthesis but tend to use different inputs.
+
+Common inputs include:
+- Parsed code corpuses
```
Comment (cyrus-) [inline src/programming-assistants.rst:81]: You should distinguish between training inputs, which come from code corpuses, from synthesis inputs, which come from the current editor state.

File: src/programming-assistants.rst [hunk 14]
```diff
@@ -1,23 +1,179 @@
...
+
 Program Repair
 ==============
 
+**Program Repair** refers to the process of automatic repairing of a code corpus that contains syntactical and/or logical errors. It is intended to speed up the time and effort exhausted to debug a software project. Debugging can consume a significant amount of time the larger or more complex the project is. Not only does the root cause of an issue have to be found but the bug itself has to be fixed making the whole process very tedious. Some common techniques used in Program Repair are statistical fault localization and component-based program synthesis.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Program repair is about fixing type errors or run-time errors, manifested as exceptions or failed tests. Syntax errors and logical errors that don't lead to type errors or exceptions are not typically handled by program repair systems.

File: src/programming-assistants.rst [hunk 15]
```diff
@@ -1,23 +1,179 @@
...
+Software engineers at Microsoft are currently developing and integrating a refactoring assistant that recognizes editing patterns in source code and suggest other edits :cite:`miltner2019fly`.
+
+Consider a corpus of code where every line is similar like the following:
+.. code-block:: python
+  var1 = ExampleFunction(param1)
+  var2 = ExampleFunction(param2)
+  var3 = ExampleFunction(param3)
+  var4 = ExampleFunction(param4)
+
+A new function is then written to initialize the created variables so the code must be changed to call the new function:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: This isn't the best example since its just a simple renaming. For describing Blue Pencil, maybe use the example from the video or just link to that blog post.

**Review by cyrus- — CHANGES_REQUESTED** [review]
topic coverage is good, but the format isn't quite right and some of the descriptions need some work -- see comments below. it should have bibliography items within the text, and each one should have a short description associated with it. please address these issues in your second draft, due on Thursday.

**Comment by abhi-ganta** [comment]
2nd draft of compendium

File: src/programming-assistants.rst [hunk 16]
```diff
@@ -1,23 +1,179 @@
-.. :Authors: - Cyrus Omar
+.. :Authors: - Cyrus Omar, Abhitej Ganta
 
 .. title:: Programming Assistants
 
 Overview
 ========
 
-TODO: Overview
+**Programming assistants** are tools integrated into development environments that enable more effective programming by allowing the developer to focus less on the syntactical and logical problems occured during development and more on the big picture of projects. Why do we need programming assistants? As more complex and time-consuming projects arise, assistants can significantly reduce the time and effort a programmer exhausts during development, while also reducing the probability of errors occuring.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: This intro sentence describes the high-level benefit of programming assistants, but you do not define the term itself. I'd rephrase this to start with a definition of programming assistants (e.g. "Programming assistants operate in concert with programmers to automate away many tedious programming tasks, including code insertion, refactoring, .....") then continue to the benefits to programmers in the rest of the introduction.

File: src/programming-assistants.rst [hunk 17]
```diff
@@ -1,23 +1,179 @@
...
 
-Refactoring
-===========
+Programming assistants can be highly sophisticated themselves, with most using advanced artificial intelligence, machine learning and natural language processing models. However, most models can be represented by the following diagram:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: I wouldn't say that "most" use AI/ML/NLP. Maybe say "some" instead.

File: src/programming-assistants.rst [hunk 18]
```diff
@@ -1,23 +1,179 @@
...
+
+.. image:: Ml_model.png
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Change "new data" to "test data" or "interactive inputs" or something like that.

Comment (cyrus-) [reply src/programming-assistants.rst]: Also maybe include "semantic context" as input to "prediction"? To emphasize that it's not just black box ML stuff

File: src/programming-assistants.rst (hunk 17 above)
Comment (cyrus-) [reply src/programming-assistants.rst]: Instead of "However, most models" -> "Most systems"

File: src/programming-assistants.rst [hunk 19]
```diff
@@ -1,23 +1,179 @@
...
+
+where the input can be range from syntactical structure (abstract syntax trees, parsed text, edit action history) to program runtime data. The chosen model is done so based on accuracy and runtime comparisons between various hypothesized models. Even after the model is chosen, it is further evaluated and tuned. Once a model is chosen and trained on input data, new data is introduced for prediction and the results are published. Models can output various types of predictions including: single hole and multiple hole fillings in a program and arbitrary edits throughout a program. The main areas of research for programming assistants are discussed further below.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: * "to program runtime data" -- mostly the predictions are about edit-time stuff, not run-time stuff
* "The chosen model is done so" -- grammar

File: src/programming-assistants.rst (hunk 18 above)
Comment (cyrus-) [reply src/programming-assistants.rst]: Maybe also change "prediction" to "suggestions"

File: src/programming-assistants.rst [hunk 20]
```diff
@@ -1,23 +1,179 @@
...
 
 Program Synthesis
 =================
 
+**Program Synthesis** refers to the process of code synthesis based on a programmer's specifications. It is intended to serve as the ultimate abstraction and allows the assistant to implement instructions typically containing a mixture of a natual language and programmatic syntax. This area of research originates from the idea that computers are very good at following explicit instructions, but humans are not necessarily good at writing those instructions. Wouldn't it be more efficient to tell a computer what the developer wants it to do and to leave the implementation details to the computer? The main promise of Program Synthesis is to automate the trivial parts of programming, allowing the developer to focus on the big picture.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: "serve as the ultimate abstraction" -- this strikes me as more marketing terminology than a technical term

File: src/programming-assistants.rst [hunk 21]
```diff
@@ -1,23 +1,179 @@
...
+
+Before discussing how synthesis occurs, it is important to discuss the various ways to specify program behavior. Most programming languages are not helpful for telling a program synthesizer what a developer wants it to do. Therefore, there must be a way to provide a specification of what the program should do. There are many styles in place including:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: "Most programming languages are not helpful for telling a program synthesizer what a developer wants it to do. Therefore, there must be a way to provide a specification of what the program should do." -- not necessary, you can just remove this

File: src/programming-assistants.rst [hunk 22]
```diff
@@ -1,23 +1,179 @@
...
+
+1. Complete Formal Specification using a formula (e.g. `∀x. P(x) = x*2` for a program that doubles the input)
+2. Input/Output pairs that serve as examples of what the program should be doing (e.g. (2,4), (4,8) for the program that doubles)
```
Comment (cyrus-) [inline src/programming-assistants.rst]: "pairs" -> "examples"

File: src/programming-assistants.rst [hunk 23]
```diff
@@ -1,23 +1,179 @@
...
+3. Demonstrations of how the program should compute the output (Very similar to #2 but tends to have some intermediate steps as well)
+
+The two main types of Program Synthesis are deductive and inductive. **Deductive Program Synthesis** uses complete formal specifications while **Inductive Program Synthesis** uses less formal specifications and relies primarily on iterative search techniques to find implementations
```
Comment (cyrus-) [inline src/programming-assistants.rst]: end in period

File: src/programming-assistants.rst [hunk 24]
```diff
@@ -1,23 +1,179 @@
...
+Deductive Program Synthesis is focused on an approach in which the derivation task is regarded as a problem of proving a mathematical theorem :cite:`manna1992fundamentals`. Manna and Waldinger discuss that in order to write a program that meets the specifications, a proof of existence must be conducted showing that for any input, the output satisfies the specified conditions. In order for the proof to be sufficiently constructive, upon finding the existence of an output, it is required to state the computational means used to determine the existence. This is then used as a basis for the program that is eventually created. Manna and Waldinger continue to discuss specifications, proving of theorems and deduction rules for the remainder of the paper.
+
+Inductive Program Synthesis (IPS) has been a more popular area of research in the last decade. This is mainly due to the realization that most are not familiar with complicated mathematical theorems and prefer to synthesize using less formal specifications. There are many different variations of IPS as there are many different iterative approaches, but most follow the same overall architecture containing a synthesizer and a verifier:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: don't make statements about relative popularity of areas, people get offended :)

File: src/programming-assistants.rst [hunk 25]
```diff
@@ -1,23 +1,179 @@
...
+
+.. image:: program_synthesis_model.png
+
+A synthesizer produces a 'candidate program' based on the specifications provided and sends the program to a verifier to validate that the program fully satisfies the specifications. If the verifier successfully validates the program, no further action is taken. However, if the verifier determines that the candidate is not satisfactory, it provides feedback to the synthesizer to be used to guide the development of a new candidate program. This process is repeated until the candidate program passes all of the specification.
+
+Gvero and Kuncak have a proposed a new code assistance tool, anyCode, for integrated development environments (IDEs) that produces a list of syntheses based on user-entered free-form queries :cite:`gvero2015synthesizing`. The queries contain a mixture of English and Java syntax and produce expressions that account for the correct syntax, types and scoping rules of Java.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Put this in the summary for the corresponding bib item below (it's okay if it is long)

File: src/programming-assistants.rst [hunk 26]
```diff
@@ -1,23 +1,179 @@
...
+anyCode was evaluated with 45 examples where the effectiveness criteria was that it was able to successfully synthesize expected expressions and to list them among the top 10 solutions. The two models used to build the tool, unigram and PCFG, were evaluated separately and together. Using only the unigram model, the expected expression was only in the top 10 27% of the time and 18% using the PCFG model.
+
+However, when both models were used, the expected expression was in the top 10 82% of time and was the #1 option 44% of the time. More implementation details of the two models can be found in the paper. anyCode was given an average runtime of 60 ms to generate the top 10 expressions, proving that it can effectively and efficiently synthesize expressions.
+
+.. container:: bib-item
+
+  .. bibliography:: programming-assistants.bib
+    :filter: key == 'manna1992fundamentals'
+
+  This paper explains the foundations and fundamentals of deductive program synthesis detailing the mathematical approach behind it.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Create separate sub-sections for deductive and inductive synthesis, and include this in the deductive section along with your introduction above.

File: src/programming-assistants.rst [hunk 27]
```diff
@@ -1,23 +1,179 @@
...
 Code Completion
 ---------------
 
+**Code Completion** systems are simple program synthesis systems that generate code snippets for insertion at the cursor, whereas other program synthesis systems may fill multiple holes in a code corpus. It is intended to speed up the process of coding applications by reducing typos and other mistakes. Code Completion assistants serves as convenient way to access function syntax and descriptions, significantly reducing the amount of memorization required.
+
+Common techniques to achieve this include auto completion popups while typing, querying parameters of functions, querying hints related to syntax errors. The models used in Code Completion are very similar to those used in Program Synthesis but tend to use different inputs.
+
+Common training inputs include:
+- Parsed code corpuses
+- Abstract syntax trees
```
Comment (cyrus-) [inline src/programming-assistants.rst:82]: Add typing context

File: src/programming-assistants.rst [hunk 28]
```diff
@@ -1,23 +1,179 @@
...
+
+Most of the popular IDEs (Visual Studio, Eclipse, IntelliSense, IntelliJ IDEA) contain some form of basic code completion but many new research projects have started picking up.
+
+Raychev, Vechev and Yahav addressed the code completion problem using APIs. Unlike most code completion tools, their work synthesizes completions for holes in code with most likely method calls :cite:`raychev2014code`. Take the following simple machine learning example:
```
Comment (cyrus-) [inline src/programming-assistants.rst]: These should also be in the paper summaries (see comment above)

File: src/programming-assistants.rst [hunk 29]
```diff
@@ -1,23 +1,179 @@
...
+3. Holes as sequences - The tool is able to recognize if multiple method invocations are missing in a given hole and fills in the missing.
+4. New fused completions - The tool is able to make completions on sequences and methods it has not been trained on before.
+
+The evaluation of the tools proved it to be very fast and effective. Out of the 84 examples tested on, the desired completion appeared in the top 3 results in 90% of the cases. More details on the implementation of the model and the evaluation metrics used can be found in the paper :cite:`raychev2014code`.
+
+Pythia, an AI-assisted code completion system, is another Microsoft research project that has since been deployed to Visual Studio IntelliCode :cite:`svyatkovskiy2019pythia`. Pythia not only suggests method completions but also search for relevant APIs for suggestion. Pythia also uses abstract syntax trees as input to train large-scale deep learning models and attempts to predict completions within 100 ms. To compare the perform of the Pythia model, the researchers also generated models using the frequency approach and Markov-Chain approach.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: same

File: src/programming-assistants.rst [hunk 30]
```diff
@@ -1,23 +1,179 @@
...
+  This paper demonstrates the use of APIs to generate code completions for holes in corpuses given a query character.
+
+.. container:: bib-item
+
+  .. bibliography:: programming-assistants.bib
+    :filter: key == 'svyatkovskiy2019pythia'
+
+  This paper demonstrates the use of abstract syntax trees to train a deep learning model for code completion suggestions.
+
+Refactoring
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Let's generalize this section "Refactoring and Program Transformation"

File: src/programming-assistants.rst [hunk 31]
```diff
@@ -1,23 +1,179 @@
...
+===========
+
+**Refactoring** refers to the process of restructuring existing code without changing the program's behavior and is a large area of interest in programming assistant research. It is intended to improve the nonfunctional attributes of a program. Typically, code refactoring applies small changes to various areas of a program and most development environments have already integrated support for basic refactoring. Detecting problems early and refactoring source code can save developers time and effort debugging in the future.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Include mention of general program transformations here too, since not all suggested transformations are refactorings.

File: src/programming-assistants.rst [hunk 32]
```diff
@@ -1,23 +1,179 @@
...
+
+The main benefits of refactoring can be categorized into either **maintainability**: Fixing issues is very easy because the code is easy to read and easy to understand, or **extensibility**: Extending the capabilities of the application is very easy because the codes uses recognizable design patterns and is flexible.
+
+(TODO: Suggesting Program Edits)
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Finish this section up for your final draft.

File: src/programming-assistants.rst [hunk 33]
```diff
@@ -1,23 +1,179 @@
...
+
+Software engineers at Microsoft are currently developing and integrating a refactoring assistant that recognizes editing patterns in source code and suggest other edits :cite:`miltner2019fly`.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Put into paper summary (see comments above)

File: src/programming-assistants.rst [hunk 34]
```diff
@@ -1,23 +1,179 @@
...
+
 Program Repair
 ==============
 
+**Program Repair** refers to the process of automatic repairing of a code corpus that contains type and/or runtime errors, manifested as exceptions or failed tests. It is intended to speed up the time and effort exhausted to debug a software project. Debugging can consume a significant amount of time the larger or more complex the project is. Not only does the root cause of an issue have to be found but the bug itself has to be fixed making the whole process very tedious. Some common techniques used in Program Repair are statistical fault localization and component-based program synthesis.
+
+A very common approach for Program Repair is using **Genetic Programming**, a computational method inspired by biological evolution which evolves computer programs tailored to a specific task. Researchers at CMU are doing just this by combining program analysis methods with evolutionary computation to automatically repair bugs :cite:`weimer2010automatic`. The key feature about the research was that it did not rely on formal specifications, allowing it to be more flexible to a larger range of software.
```
Comment (cyrus-) [inline src/programming-assistants.rst]: Put these into paper summaries too (see above)

**Review by cyrus- — CHANGES_REQUESTED** [review]
nice job overall. see comments for final draft.

File: src/programming-assistants.bib [hunk 35]
```diff
@@ -0,0 +1,68 @@
+@inproceedings{gvero2015synthesizing,
+  title={Synthesizing Java expressions from free-form queries},
```
Comment (cyrus-) [inline src/programming-assistants.bib]: need to put proper nouns inside `{}` so that they show up correctly (e.g. `{Java}`)

File: src/programming-assistants.bib [hunk 36]
```diff
@@ -0,0 +1,68 @@
...
+  author={Gvero, Tihomir and Kuncak, Viktor},
+  booktitle={Acm Sigplan Notices},
```
Comment (cyrus-) [inline src/programming-assistants.bib]: don't cite SIGPLAN notices (it just republishes conference papers), cite the original conference

File: src/programming-assistants.bib [hunk 37]
```diff
@@ -0,0 +1,68 @@
...
+  number={10},
+  pages={416--432},
+  year={2015},
+  organization={ACM}
+}
+
+@inproceedings{svyatkovskiy2019pythia,
+  title={Pythia: AI-assisted Code Completion System},
+  author={Svyatkovskiy, Alexey and Zhao, Ying and Fu, Shengyu and Sundaresan, Neel},
+  booktitle={Proceedings of the 25th ACM SIGKDD International Conference on Knowledge Discovery \& Data Mining},
```
Comment (cyrus-) [inline src/programming-assistants.bib]: Change to just "KDD"

File: src/programming-assistants.bib [hunk 38]
```diff
@@ -0,0 +1,68 @@
...
+  pages={2727--2735},
+  year={2019},
+  organization={ACM}
+}
+
+@inproceedings{raychev2014code,
+  title={Code completion with statistical language models},
+  author={Raychev, Veselin and Vechev, Martin and Yahav, Eran},
+  booktitle={Acm Sigplan Notices},
```
Comment (cyrus-) [inline src/programming-assistants.bib]: no sigplan notices

File: src/programming-assistants.bib [hunk 39]
```diff
@@ -0,0 +1,68 @@
...
+  number={5},
+  pages={109--116},
+  year={2010},
+  publisher={ACM}
+}
+
+@inproceedings{nguyen2013semfix,
+  title={Semfix: Program repair via semantic analysis},
+  author={Nguyen, Hoang Duong Thien and Qi, Dawei and Roychoudhury, Abhik and Chandra, Satish},
+  booktitle={2013 35th International Conference on Software Engineering (ICSE)},
```
Comment (cyrus-) [inline src/programming-assistants.bib]: just ICSE

File: src/programming-assistants.rst (hunk 25 above)
Comment (abhi-ganta) [reply src/programming-assistants.rst]: Should there be no text besides the citation then?
